File: composition/Scala/composition-pipeline/pipeline/src/main/scala/common/Pipeline.scala
package common

import scala.collection.mutable.ListBuffer

object Pipeline {
  case class Result[TData](error: String = "", data: TData = None) {
    def success() = error == ""
  }

  def runAll[TData](data: TData, funcs: ((TData) => Result[TData])*): Result[TData] = {
    val errors = new ListBuffer[String]()
    var currentData = data
    funcs.foreach(f => {
      val result = f(currentData)
      if (!result.success())
        errors += result.error
      currentData = result.data
    })
    Result(errors.toList.mkString("<br/>"), currentData)
  }

  def runUntilFirstFault[TData](data: TData, funcs: ((TData) => Result[TData])*): Result[TData] = {
    var result: Result[TData] = null
    funcs.foreach(f => {
      result = f(data)
      if (!result.success())
        return result
    })
    Result(data = result.data)
  }
}

File: composition/JavaScript/pipeline/src/pipeline.js
import { Result } from "./result";
import _ from 'lodash';

export function runAll(data, funcs) {
    const errors = [];
    let currentData = data;
    _.forEach(funcs, (f) => {
        const result = f(currentData);
        currentData = result.getData();
        if (!result.success()) errors.push(result.getError());
    });
    return new Result(errors.join("<br/>"), currentData);
}

export function runUntilFirstFault(data, funcs) {
    function hasError(data, funcs) {
        if (funcs.length === 0) return "";
        const result = funcs[0](data);
        if (!result.success()) return result.getError();
        funcs.shift();
        return hasError(result.getData(), funcs);
    }
    
    return new Result(hasError(data, funcs));
}